drivers/can/can_sam.c

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2021 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/can.h>
+#include <soc.h>
+#include <kernel.h>
+#include "can_sam.h"
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
+#define DT_DRV_COMPAT atmel_sam_can
+
+int can_sam_get_core_clock(const struct device *dev, uint32_t *rate)
+{
+
+	*rate = SOC_ATMEL_SAM_MCK_FREQ_HZ / (CONFIG_CAN_SAM_CKDIV + 1);
+	return 0;
+}
+
+void can_sam_set_state_change_callback(const struct device *dev,
+				       can_state_change_callback_t isr)
+{
+	struct can_sam_data *data = DEV_DATA(dev);
```

Comment:
Please, don't use `DEV_DATA/DEV_CONFIG` macros in new code, ref. https://github.com/zephyrproject-rtos/zephyr/issues/37616

Reply:
Done.